Add one_qubit_decomposition and zxz_decomposition functions

Newtech66 opened the pull request:
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [x] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
A unified function for implementing one qubit unitary decompositions is required. Additionally, support for ZXZ decomposition is required.

**Description of the Change:**
Adds functions one_qubit_decomposition and zxz_decomposition to single_qubit_unitary.py implementing the above. Also fixes a previously existing bug in the implementation of xyx_decomposition which made it return incorrect results for certain unitaries.

**Benefits:**
One function for different decompositions, so it is more convenient.

**Possible Drawbacks:**
None.

**Related GitHub Issues:**
Fixes #4121

rmoyard commented:
Hi @Newtech66 ! Let me know when I can take a look or if you have some questions, thanks

Newtech66 commented:
Hi! There are some questions in the thread of the original issue that need to be addressed. I'll just summarise my work:

- `zyz_decomposition` returned a `Rot` gate before. Now it returns a sequence of `ZYZ` gates similarly to `xyx_decomposition`.
- If the unitary is diagonal, `zyz_decomposition` returned a single `RZ` gate. I retained this behaviour because it's better to return a shorter circuit where reasonably possible. Unfortunately, this breaks the tests for `two_qubit_decomposition` because they test the number of obtained gates which is no longer constant (earlier it would have been 1). For now, I worked around this by disabling the checks for number of obtained gates, which was enough.
- I added `zxz_decomposition`. It works the same way as `zyz_decomposition`.
- I added `one_qubit_decomposition`. It has the same signature as mentioned in the original issue.
- All of the above functions now support `return_global_phase` as a parameter which returns the global phase if it is `True`.
- All of the angles are now explicitly constrained within $[-\pi, \pi]$.
- The individual decomposition functions now return gates in the order they should be applied to recover the unitary $U$. For example, if $U=RZ(\omega)RX(\theta)RZ(\phi)$, then the output is `[RZ(phi), RX(theta), RZ(omega)]`. Earlier, the behaviour was instead `[RZ(omega), RX(theta), RZ(phi)]`.

Question:

- Should batched unitaries be supported? The implementation and tests of `zyz_decomposition` explicitly support batched unitaries, but `xyx_decomposition` does not make any mention of it. The batched unitaries tests for `zyz_decomposition` are broken as well, so I disabled them for now.

Newtech66 commented:
Actually, it seems lots of tests which check the length of circuits assuming `zyz_decomposition` returns a fixed number of gates are now breaking. Maybe we should remove the case of returning a single `RZ` when the unitary is diagonal?

rmoyard commented:
Hi @Newtech66!

1) You can add batching support but it is not mandatory for unitary hack.
2) Yes you can make the function private and document the dispatcher.

I will take a look later today or tomorrow.

Thanks!

rmoyard commented on pennylane/transforms/decompositions/single_qubit_unitary.py:
```diff
@@ -156,26 +152,145 @@ def xyx_decomposition(U, wire, return_global_phase=False):
     return Operations
+
+
+def one_qubit_decomposition(U, rotations, wire, return_global_phase=False):
```
Could you add ZYZ as default?

Newtech66 replied:
`wire` does not have a default argument, so if I added a default argument for `rotations`, I can't have a non-default argument following a default argument. What's the best way to deal with this?

rmoyard commented on pennylane/transforms/decompositions/two_qubit_unitary.py:
```diff
@@ -19,7 +19,7 @@
 from pennylane import math
 
-from .single_qubit_unitary import zyz_decomposition
+from .single_qubit_unitary import _zyz_decomposition
```
Could you use the public function with zyz rotations?

rmoyard commented on pennylane/transforms/decompositions/single_qubit_unitary.py:
```diff
@@ -43,100 +43,96 @@ def _convert_to_su2(U, return_global_phase=False):
-    return [qml.Rot(phis, thetas, omegas, wires=wire)]
+    phis, thetas, omegas, alphas = map(math.squeeze, [phis, thetas, omegas, alphas])
+
+    Operations = [qml.RZ(phis, wire), qml.RY(thetas, wire), qml.RZ(omegas, wire)]
```
```suggestion
    operations = [qml.RZ(phis, wire), qml.RY(thetas, wire), qml.RZ(omegas, wire)]
```
We do not user upper case at the start of a variable

Newtech66 replied:
Fixed in next commit.

rmoyard commented on pennylane/transforms/decompositions/single_qubit_unitary.py:
```diff
@@ -145,9 +141,9 @@ def xyx_decomposition(U, wire, return_global_phase=False):
-    [RX(array(0.45246584), wires=[0]),
-    RY(array(1.39749741), wires=[0]),
-    RX(array(-1.72101925), wires=[0]),
+    [RX(tensor(-1.72101925, requires_grad=True), wires=[0]),
```
Why the angle are different than previously?

Newtech66 replied:
Notice that the order is just flipped. The angles are the same. Previously the gates were returned in matrix multiplication order (so multiplying left to right gives the unitary). Now, I changed it so that the gates are returned in the order they must be applied to the qubit (in other words, the order in which they should appear in the circuit). So multiplying right to left gives the unitary now.

rmoyard replied:
Thank you it is a very good catch!

rmoyard commented on pennylane/transforms/decompositions/single_qubit_unitary.py:
```diff
@@ -156,26 +152,145 @@ def xyx_decomposition(U, wire, return_global_phase=False):
+
+    # Wrap angles to range [-\pi, \pi]
+    lams = (lams + qml.numpy.pi) % (2 * qml.numpy.pi) - qml.numpy.pi
+    phis = (phis + qml.numpy.pi) % (2 * qml.numpy.pi) - qml.numpy.pi
```
Is it because you simplify the angle that th logic is slightly modify?

Newtech66 replied:
Yes, acquiring the angles is the same as before, just added two extra lines to normalize them to $[-\pi, \pi]$.

Newtech66 commented:
Hi @rmoyard, I think it's ready for reviewing. Please take a look.

rmoyard reviewed (commented):
You are on a very good way! I have left some initial thoughts. Thanks!

Newtech66 commented:
I pushed some commits to fix the issues, why aren't they showing up here? They are showing up on my branch though.

rmoyard commented on pennylane/transforms/decompositions/single_qubit_unitary.py:
```diff
@@ -43,100 +43,96 @@ def _convert_to_su2(U, return_global_phase=False):
     return (U_SU2, exp_angles) if return_global_phase else U_SU2
 
 
-def zyz_decomposition(U, wire):
```
Don't feel obligated to do it because it is not required for unitary hack, but would it be possible to copy paste the previous function and create: `rot_decomposition with the keyword rot` because that what was zyz in fact?

Newtech66 replied:
What exactly do I need to do? Copy paste the old function verbatim and add it to `supported_rotations` with key `rot`? Also, the original function didn't have support for returning global phase. I might also have to add tests for code coverage.

Which of the above do you want me to do?

Newtech66 replied:
I added `_rot_decomposition` (accessible using `rot` in one_qubit_decomposition`) and ignored if `return_global_phase` was passed as `True`. Also added tests for it. Didn't add it as a supported option in documentation for `one_qubit_decomposition` though.

rmoyard replied:
Thanks that looks good to me!

rmoyard commented on tests/transforms/test_decompositions.py:
```diff
@@ -754,7 +987,7 @@ def test_two_qubit_decomposition_3_cnots(self, U, wires):
 
         obtained_decomposition = two_qubit_decomposition(U, wires=wires)
-        assert len(obtained_decomposition) == 10
+        assert len(obtained_decomposition) == 18
```
Just to make sure why those changes are there. It is because previsouly we would get Rot instead of the decomposition?

Newtech66 replied:
Yes. So now 3 gates are being returned instead of 1.

rmoyard reviewed (commented):
Great job!

```
/github/workspace/pennylane/transforms/decompositions/single_qubit_unitary.py:docstring of pennylane.transforms.decompositions.single_qubit_unitary.one_qubit_decomposition:16: WARNING: Inline interpreted text or phrase reference start-string without end-string.
looking for now-outdated files... none found
/github/workspace/pennylane/transforms/decompositions/single_qubit_unitary.py:docstring of pennylane.transforms.decompositions.single_qubit_unitary.one_qubit_decomposition:17: WARNING: Inline interpreted text or phrase reference start-string without end-string.
```
There is still a small mistake in the sphinx build for documentation otherwise it looks great! One suggestion is to keep the old zyz decomposition and rename it rot but that's not obligatory for unitary hack. 

One thing that I will do is to deprecate the old one qubit function and keep them around for a bit and internally it will call the private one. We might have some conflict issues after that, I keep you updated!

rmoyard reviewed (commented):
Thanks @Newtech66 amazing PR!

Two last things to do: 
1) add the changelog entry with your name as a contributor.
2) I deprecated those two functions in a separate PR https://github.com/PennyLaneAI/pennylane/pull/4230 . Two options either you wait for it to be merged and then solve the conflicts or you copy paste the changes. I am happy with both solutions, do what is more convenient for you.

Thanks a lot for contributing!

Newtech66 commented:
@rmoyard I'll wait for the other PR to be merged then I'll make the finishing contributions. Glad I was able to help!

rmoyard commented:
Hi @Newtech66 the deprecation PR is merged! You can go ahead. Let me know if you have any questions

Newtech66 commented:
I added my name to the contributors list in the changelog, but I'm not sure what to fill in for the rest...

Newtech66 commented:
@rmoyard Ready for review!

dwierichs commented on pennylane/transforms/decompositions/single_qubit_unitary.py:
```diff
@@ -169,22 +169,96 @@ def _zyz_decomposition(U, wire):
 
+def _zyz_decomposition(U, wire, return_global_phase=False):
+    r"""Compute the decomposition of a single-qubit matrix :math:`U` in terms
+    of elementary operations, as a product of X and Y rotations in the form
```
```suggestion
    of elementary operations, as a product of Z and Y rotations in the form
```

dwierichs commented on pennylane/transforms/decompositions/single_qubit_unitary.py:
```diff
@@ -169,22 +169,96 @@ def _zyz_decomposition(U, wire):
-        another ``RX`` gate, which when applied in the order of appearance in the list is equivalent
-        to the unitary :math:`U` up to global phase. If `return_global_phase=True`,
+        another ``RX`` gate, which when applied in the order of appearance in the list is
+        equivalent to the unitary :math:`U` up to global phase. If `return_global_phase=True`,
```
```suggestion
        equivalent to the unitary :math:`U` up to a global phase. If `return_global_phase=True`,
```

dwierichs commented on pennylane/transforms/decompositions/single_qubit_unitary.py:
```diff
@@ -240,26 +314,147 @@ def _xyx_decomposition(U, wire, return_global_phase=False):
-        theta = 2 * math.arccos(math.real(U_det1[:, 1, 1]) / (math.cos(lam_plus_phi) + EPS))
-    else:
-        theta = 2 * math.arccos(-math.imag(U_det1[:, 0, 1]) / (math.sin(lam_plus_phi) + EPS))
+    thetas = math.array(
```
Not required, but one could solve this with a `math.where` call, I think :)

Newtech66 replied:
I didn't know about this function, I'll use it.

dwierichs commented on pennylane/transforms/decompositions/single_qubit_unitary.py:
```diff
@@ -240,26 +314,147 @@ def _xyx_decomposition(U, wire, return_global_phase=False):
+    phis, thetas, lams, gammas = map(math.squeeze, [phis, thetas, lams, gammas])
 
-    Operations = [qml.RX(phi, wire), qml.RY(theta, wire), qml.RX(lam, wire)]
+    operations = [qml.RX(lams, wire), qml.RY(thetas, wire), qml.RX(phis, wire)]
```
:+1:

dwierichs commented on pennylane/transforms/decompositions/single_qubit_unitary.py:
```diff
@@ -240,26 +314,147 @@ def _xyx_decomposition(U, wire, return_global_phase=False):
+    Returns:
+        list[Operation]: Returns a list of gates, an ``RZ``, an ``RX`` and
+        another ``RZ`` gate, which when applied in the order of appearance in the list is
+        equivalent to the unitary :math:`U` up to global phase. If `return_global_phase=True`,
```
```suggestion
        equivalent to the unitary :math:`U` up to a global phase. If `return_global_phase=True`,
```

dwierichs commented on pennylane/transforms/decompositions/single_qubit_unitary.py:
```diff
@@ -169,22 +169,96 @@ def _zyz_decomposition(U, wire):
+
+    operations = [qml.RZ(phis, wire), qml.RY(thetas, wire), qml.RZ(omegas, wire)]
+    if return_global_phase:
+        operations += [qml.s_prod(math.exp(1j * alphas), qml.Identity(wire))]
```
```suggestion
        operations.append(qml.s_prod(math.exp(1j * alphas), qml.Identity(wire)))
```

dwierichs commented on pennylane/transforms/decompositions/single_qubit_unitary.py:
```diff
@@ -240,26 +314,147 @@ def _xyx_decomposition(U, wire, return_global_phase=False):
+    operations = [qml.RX(lams, wire), qml.RY(thetas, wire), qml.RX(phis, wire)]
     if return_global_phase:
-        Operations += [qml.s_prod(math.exp(1j * gamma)[0], qml.Identity(wire))]
+        operations += [qml.s_prod(math.exp(1j * gammas), qml.Identity(wire))]
```
```suggestion
        operations.append(qml.s_prod(math.exp(1j * gammas), qml.Identity(wire)))
```

dwierichs commented on pennylane/transforms/decompositions/single_qubit_unitary.py:
```diff
@@ -240,26 +314,147 @@ def _xyx_decomposition(U, wire, return_global_phase=False):
+    # Return gates in the order they will be applied on the qubit
+    operations = [qml.RZ(psis, wire), qml.RX(thetas, wire), qml.RZ(phis, wire)]
+    if return_global_phase:
+        operations += [qml.s_prod(math.exp(1j * alphas), qml.Identity(wire))]
```
```suggestion
        operations.append(qml.s_prod(math.exp(1j * alphas), qml.Identity(wire)))
```

dwierichs commented on pennylane/transforms/decompositions/single_qubit_unitary.py:
```diff
@@ -240,26 +314,147 @@ def _xyx_decomposition(U, wire, return_global_phase=False):
+        U (tensor): A :math:`2 \times 2` unitary matrix.
+        wire (Union[Wires, Sequence[int] or int]): The wire on which to apply the operation.
+        rotations (str): A string defining the sequence of rotations to decompose :math:`U` into.
+        return_global_phase (bool): Whether to return the global phase as a `qml.s_prod` between `exp(1j)*alpha`
+            and `qml.Identity` as the last element of the returned list of operations.
```
```suggestion
        return_global_phase (bool): Whether to return the global phase as a ``qml.s_prod`` between ``exp(1j)*alpha``
            and ``qml.Identity`` as the last element of the returned list of operations.
```

dwierichs commented on pennylane/transforms/decompositions/single_qubit_unitary.py:
```diff
@@ -240,26 +314,147 @@ def _xyx_decomposition(U, wire, return_global_phase=False):
+
+    Returns:
+        list[Operation]: Returns a list of gates which when applied in the order of appearance in
+        the list is equivalent to the unitary :math:`U` up to global phase. If `return_global_phase=True`,
```
```suggestion
        the list is equivalent to the unitary :math:`U` up to a global phase. If `return_global_phase=True`,
```

dwierichs commented on pennylane/transforms/decompositions/single_qubit_unitary.py:
```diff
@@ -240,26 +314,147 @@ def _xyx_decomposition(U, wire, return_global_phase=False):
+            return supported_rotations[rotations](U, wire)
+        return supported_rotations[rotations](U, wire, return_global_phase)
+
+    raise ValueError("Value passed to rotations is either invalid or currently unsupported.")
```
```suggestion
    raise ValueError(f"Value {rotations} passed to rotations is either invalid or currently unsupported.")
```

Newtech66 commented on doc/releases/changelog-dev.md:
```diff
@@ -325,6 +337,7 @@ Christina Lee,
 Tristan Nemoz,
+Mainak Roy,
 Mudit Pandey,
 Borja Requena,
```
Yep, done.

Newtech66 commented on doc/releases/changelog-dev.md:
```diff
@@ -310,6 +319,9 @@
 * Fixes a bug where `op = qml.qsvt()` was incorrect up to a global phase when using `convention="Wx""` and `qml.matrix(op)`.
   [(#4214)](https://github.com/PennyLaneAI/pennylane/pull/4214)
 
+* Fixed buggy calculation of angle in `xyx_decomposition` causing it to give incorrect decomposition of certain unitaries.
```
Added more details to changelog.

dwierichs commented on pennylane/transforms/decompositions/single_qubit_unitary.py:
```diff
@@ -240,26 +314,143 @@ def _xyx_decomposition(U, wire, return_global_phase=False):
-    else:
-        theta = 2 * math.arccos(-math.imag(U_det1[:, 0, 1]) / (math.sin(lam_plus_phi) + EPS))
+    thetas = math.where(
+        math.isclose(math.sin(lams_plus_phis), 0.0 * math.ones_like(lams_plus_phis)),
```
```suggestion
        math.isclose(math.sin(lams_plus_phis), math.zeros_like(lams_plus_phis)),
```

dwierichs commented on pennylane/transforms/decompositions/single_qubit_unitary.py:
```diff
@@ -240,26 +314,143 @@ def _xyx_decomposition(U, wire, return_global_phase=False):
+    thetas = math.where(
+        math.isclose(math.sin(lams_plus_phis), 0.0 * math.ones_like(lams_plus_phis)),
+        2 * math.arccos(math.real(U_det1[:, 1, 1]) / (math.cos(lams_plus_phis) + EPS)),
+        2 * math.arccos(-math.imag(U_det1[:, 0, 1]) / (math.sin(lams_plus_phis) + EPS)),
```
Excellent, thank you for going the extra mile! :heart_eyes:

dwierichs commented on pennylane/transforms/decompositions/single_qubit_unitary.py:
```diff
@@ -240,26 +314,143 @@ def _xyx_decomposition(U, wire, return_global_phase=False):
+
+    # Conditional to avoid divide by 0 errors
+    thetas = math.where(
+        math.isclose(math.sin(phis_plus_psis), 0.0 * math.ones_like(phis_plus_psis)),
```
```suggestion
        math.isclose(math.sin(phis_plus_psis), math.zeros_like(phis_plus_psis)),
```

dwierichs reviewed (approved):
Very thorough addition of tests, great job @Newtech66 !
Couldn't find any issue or gap.
Just two tiny comments regarding `zeros_like` over `0.0 * ones_like` :)

Newtech66 commented:
All done!
